File: homescreen.go
package main

import (
	"image/color"
	"log"

	"github.com/hajimehoshi/ebiten/v2"
	"github.com/hajimehoshi/ebiten/v2/ebitenutil"
	"github.com/hajimehoshi/ebiten/v2/inpututil"
	"github.com/hajimehoshi/ebiten/v2/vector"
)

type HomeScreen struct {
	gameButtons []*Button
	retryButton *Button
}

func NewHomeScreen() *HomeScreen {
	hs := &HomeScreen{
		gameButtons: make([]*Button, 4),
	}

	// Create game selection buttons - 2x2 grid (4 games)
	buttonWidth := 280.0
	buttonHeight := 90.0
	startX := float64(screenWidth/2) - buttonWidth - 20
	startY := 250.0
	spacingX := buttonWidth + 40.0
	spacingY := buttonHeight + 20.0

	hs.gameButtons[0] = &Button{
		x:       startX,
		y:       startY,
		width:   buttonWidth,
		height:  buttonHeight,
		text:    "YAHTZEE",
		enabled: true,
	}

	hs.gameButtons[1] = &Button{
		x:       startX + spacingX,
		y:       startY,
		width:   buttonWidth,
		height:  buttonHeight,
		text:    "SANTORINI",
		enabled: true,
	}

	hs.gameButtons[2] = &Button{
		x:       startX,
		y:       startY + spacingY,
		width:   buttonWidth,
		height:  buttonHeight,
		text:    "CONNECT FOUR",
		enabled: true,
	}

	hs.gameButtons[3] = &Button{
		x:       startX + spacingX,
		y:       startY + spacingY,
		width:   buttonWidth,
		height:  buttonHeight,
		text:    "MEMORY MATCH",
		enabled: true,
	}

	// Add "Retry" button (only shown when connection fails)
	hs.retryButton = &Button{
		x:       float64(screenWidth/2) - 100,
		y:       float64(screenHeight) - 100,
		width:   200,
		height:  60,
		text:    "RETRY CONNECTION",
		enabled: true,
	}

	return hs
}

func (hs *HomeScreen) Update(gr *GameRoom) error {
	// Only show home screen if not connected
	if gr.connectionState != StateConnected {
		x, y := ebiten.CursorPosition()

		// Update retry button hover state (only when connection failed)
		if gr.connectionState == StateFailed && hs.retryButton != nil {
			hs.retryButton.hovered = hs.retryButton.Contains(x, y)
		}

		// Handle retry button click
		if inpututil.IsMouseButtonJustPressed(ebiten.MouseButtonLeft) {
			if gr.connectionState == StateFailed && hs.retryButton != nil && hs.retryButton.hovered {
				log.Println("Retry button clicked - attempting to reconnect")
				gr.TryGoOnline()
			}
		}
	}

	return nil
}

func (hs *HomeScreen) Draw(screen *ebiten.Image, gr *GameRoom) {
	// Draw background
	DrawForestBackground(screen)
	DrawKodamaSpirits(screen)
	DrawOMLogo(screen)

	// Draw title
	titleWidth := float32(500)
	titleX := float32(screenWidth/2) - titleWidth/2
	titleY := float32(80)

	vector.DrawFilledRect(screen, titleX, titleY, titleWidth, 80, color.RGBA{30, 50, 80, 255}, false)
	vector.StrokeRect(screen, titleX, titleY, titleWidth, 80, 3, color.RGBA{100, 150, 220, 255}, false)

	// Title text
	title1 := "OLIVE & MILLIE'S"
	title2 := "GAME ROOM"
	title1X := int(titleX + (titleWidth-float32(len(title1)*6))/2)
	title2X := int(titleX + (titleWidth-float32(len(title2)*6))/2)

	ebitenutil.DebugPrintAt(screen, title1, title1X, int(titleY+20))
	ebitenutil.DebugPrintAt(screen, title1, title1X+1, int(titleY+20))
	ebitenutil.DebugPrintAt(screen, title2, title2X, int(titleY+45))
	ebitenutil.DebugPrintAt(screen, title2, title2X+1, int(titleY+45))

	// Draw avatars next to title
	DrawPlayer1Avatar(screen, titleX-80, titleY+15, 1.0)
	DrawPlayer2Avatar(screen, titleX+titleWidth+30, titleY+15, 1.0)

	// Draw connection status message
	var statusText string
	var statusColor color.RGBA

	switch gr.connectionState {
	case StateConnecting:
		statusText = "Connecting to server..."
		statusColor = color.RGBA{100, 150, 220, 255}
	case StateFailed:
		statusText = "Connection Failed"
		statusColor = color.RGBA{200, 80, 80, 255}
	case StateConnected:
		statusText = "Connected!"
		statusColor = color.RGBA{80, 200, 80, 255}
	}

	// Draw status box
	statusWidth := float32(len(statusText)*6 + 40)
	statusX := float32(screenWidth/2) - statusWidth/2
	statusY := float32(400)
	vector.DrawFilledRect(screen, statusX, statusY, statusWidth, 50, color.RGBA{30, 50, 80, 200}, false)
	vector.StrokeRect(screen, statusX, statusY, statusWidth, 50, 3, statusColor, false)

	// Draw status text
	textX := int(statusX + statusWidth/2 - float32(len(statusText)*6)/2)
	ebitenutil.DebugPrintAt(screen, statusText, textX, int(statusY+17))
	ebitenutil.DebugPrintAt(screen, statusText, textX+1, int(statusY+17)) // Bold

	// Draw error message if failed
	if gr.connectionState == StateFailed && gr.connectionError != "" {
		errorX := screenWidth/2 - len(gr.connectionError)*3
		ebitenutil.DebugPrintAt(screen, gr.connectionError, errorX, int(statusY+60))
	}

	// Draw retry button if connection failed
	if gr.connectionState == StateFailed && hs.retryButton != nil {
		DrawButton(screen, hs.retryButton)
	}
	
}

func (hs *HomeScreen) drawGameButton(screen *ebiten.Image, btn *Button) {
	btnColor := color.RGBA{100, 150, 220, 255}
	borderColor := color.RGBA{70, 120, 190, 255}
	if btn.hovered {
		btnColor = color.RGBA{130, 180, 240, 255}
		borderColor = color.RGBA{100, 150, 220, 255}
	}

	// Shadow
	vector.DrawFilledRect(screen, float32(btn.x+4), float32(btn.y+4), float32(btn.width), float32(btn.height), color.RGBA{0, 0, 0, 60}, false)

	// Button background
	vector.DrawFilledRect(screen, float32(btn.x), float32(btn.y), float32(btn.width), float32(btn.height), btnColor, false)
	vector.StrokeRect(screen, float32(btn.x), float32(btn.y), float32(btn.width), float32(btn.height), 3, borderColor, false)

	// Button text
	textX := int(btn.x + btn.width/2 - float64(len(btn.text))*3)
	textY := int(btn.y + btn.height/2 - 5)
	ebitenutil.DebugPrintAt(screen, btn.text, textX, textY)
	ebitenutil.DebugPrintAt(screen, btn.text, textX+1, textY) // Bold
	ebitenutil.DebugPrintAt(screen, btn.text, textX, textY+1) // Bold
}

File: web_client/js/homescreen.js
// HomeScreen implementation
class HomeScreen {
    constructor(gameRoom) {
        this.gameRoom = gameRoom;
        this.gameButtons = [];
        this.avatarButtons = [];
        this.currentAvatar = 0;
        this.sparkles = [];
        this.lastSparkleTime = Date.now();
        this.sparkleEmitRate = 100; // milliseconds
        
        this.initializeButtons();
        this.initializeSparkles();
    }
    
    initializeButtons() {
        // Game selection buttons (2x2 grid)
        const gameTypes = [
            { name: "YAHTZEE", type: "yahtzee" },
            { name: "SANTORINI", type: "santorini" },
            { name: "CONNECT FOUR", type: "connect_four" },
            { name: "MEMORY MATCH", type: "memory" }
        ];
        
        const buttonWidth = 200;
        const buttonHeight = 50;
        const startX = SCREEN_WIDTH / 2 - 210;
        const startY = 300;
        const spacing = 20;
        
        for (let i = 0; i < gameTypes.length; i++) {
            const row = Math.floor(i / 2);
            const col = i % 2;
            const x = startX + col * (buttonWidth + spacing);
            const y = startY + row * (buttonHeight + spacing);
            
            const btn = new Button(x, y, buttonWidth, buttonHeight, gameTypes[i].name);
            btn.gameType = gameTypes[i].type;
            this.gameButtons.push(btn);
        }
        
        // Avatar selection buttons
        const avatarY = 500;
        const avatarSize = 40;
        const totalAvatars = AVATAR_TYPES.NUM_TYPES;
        const avatarSpacing = 5;
        const totalWidth = totalAvatars * avatarSize + (totalAvatars - 1) * avatarSpacing;
        const avatarStartX = (SCREEN_WIDTH - totalWidth) / 2;
        
        for (let i = 0; i < totalAvatars; i++) {
            const x = avatarStartX + i * (avatarSize + avatarSpacing);
            const btn = new Button(x, avatarY, avatarSize, avatarSize, "");
            btn.avatarType = i;
            this.avatarButtons.push(btn);
        }
    }
    
    initializeSparkles() {
        // Create initial sparkles around the logo
        for (let i = 0; i < 5; i++) {
            this.sparkles.push(this.createSparkle());
        }
    }
    
    createSparkle() {
        return {
            x: 75 + Math.random() * 60 - 30,
            y: 45 + Math.random() * 40 - 20,
            size: Math.random() * 3 + 2,
            lifetime: 0,
            maxLifetime: Math.random() * 1000 + 1000,
            speed: Math.random() * 0.5 + 0.1
        };
    }
    
    update(inputState) {
        // Update sparkles
        const now = Date.now();
        if (now - this.lastSparkleTime > this.sparkleEmitRate && this.sparkles.length < 10) {
            this.sparkles.push(this.createSparkle());
            this.lastSparkleTime = now;
        }
        
        // Update sparkle positions and lifetimes
        this.sparkles = this.sparkles.filter(sparkle => {
            sparkle.lifetime += 16; // ~60fps
            sparkle.y -= sparkle.speed;
            return sparkle.lifetime < sparkle.maxLifetime;
        });
        
        // Update button hover states
        for (const btn of this.gameButtons) {
            btn.hovered = btn.contains(inputState.mouseX, inputState.mouseY);
        }
        for (const btn of this.avatarButtons) {
            btn.hovered = btn.contains(inputState.mouseX, inputState.mouseY);
        }
        
        // Handle clicks
        if (inputState.mouseJustPressed) {
            console.log(`Click detected at (${inputState.mouseX}, ${inputState.mouseY})`);
            
            // Check logo click
            if (isLogoClicked(inputState.mouseX, inputState.mouseY)) {
                console.log("Logo clicked!");
            }
            
            // Check game button clicks
            for (const btn of this.gameButtons) {
                if (btn.contains(inputState.mouseX, inputState.mouseY)) {
                    console.log(`Game button clicked: ${btn.gameType}`);
                    this.gameRoom.selectGame(btn.gameType);
                    break;
                }
            }
            
            // Check avatar button clicks
            for (let i = 0; i < this.avatarButtons.length; i++) {
                const btn = this.avatarButtons[i];
                if (btn.contains(inputState.mouseX, inputState.mouseY)) {
                    this.currentAvatar = i;
                    this.gameRoom.setAvatar(i);
                    console.log(`Avatar selected: ${AVATAR_NAMES[i]}`);
                    break;
                }
            }
        }
    }
    
    draw(screen) {
        // Clear screen
        screen.clear();
        
        // Draw forest background
        drawForestBackground(screen);
        drawKodamaSpirits(screen);
        
        // Draw logo
        drawOMLogo(screen);
        
        // Draw sparkles
        for (const sparkle of this.sparkles) {
            const alpha = 1 - (sparkle.lifetime / sparkle.maxLifetime);
            const color = { 
                r: COLORS.SPARKLE.r, 
                g: COLORS.SPARKLE.g, 
                b: COLORS.SPARKLE.b, 
                a: Math.floor(alpha * 255) 
            };
            screen.fillCircle(sparkle.x, sparkle.y, sparkle.size, color);
        }
        
        // Draw title
        screen.drawText("Olive & Millie's Game Room", SCREEN_WIDTH / 2 - 120, 150, 
            { r: 255, g: 255, b: 255, a: 255 }, "20px monospace");
        
        // Draw connection status
        const statusText = this.gameRoom.isOnline() ? "ONLINE" : "OFFLINE";
        const statusColor = this.gameRoom.isOnline() ? 
            { r: 100, g: 255, b: 100, a: 255 } : 
            { r: 255, g: 100, b: 100, a: 255 };
        screen.drawText(statusText, SCREEN_WIDTH / 2 - 30, 180, statusColor, "16px monospace");
        
        // Draw instructions
        screen.drawText("Select Your Avatar:", SCREEN_WIDTH / 2 - 80, 480, 
            { r: 200, g: 200, b: 200, a: 255 }, "14px monospace");
        
        // Draw game buttons
        for (const btn of this.gameButtons) {
            drawButton(screen, btn);
        }
        
        // Draw avatar buttons
        for (let i = 0; i < this.avatarButtons.length; i++) {
            const btn = this.avatarButtons[i];
            
            // Highlight selected avatar
            if (i === this.currentAvatar) {
                screen.fillRect(btn.x - 2, btn.y - 2, btn.width + 4, btn.height + 4, 
                    { r: 255, g: 220, b: 100, a: 255 });
            }
            
            // Draw avatar
            this.drawAvatar(screen, i, btn.x + btn.width / 2, btn.y + btn.height / 2, 0.8);
            
            // Draw hover effect
            if (btn.hovered) {
                screen.strokeRect(btn.x - 1, btn.y - 1, btn.width + 2, btn.height + 2, 2,
                    { r: 255, g: 255, b: 255, a: 200 });
            }
        }
        
        // Draw selected avatar name
        screen.drawText(AVATAR_NAMES[this.currentAvatar], SCREEN_WIDTH / 2 - 50, 560,
            { r: 255, g: 255, b: 255, a: 255 }, "16px monospace");
    }
    
    drawAvatar(screen, avatarType, x, y, scale = 1.0) {
        // Simplified avatar drawing for web
        // In a full implementation, this would match the Go version's avatar drawing
        const size = 20 * scale;
        const color = this.getAvatarColor(avatarType);
        
        // Draw simple circle for now
        screen.fillCircle(x, y, size, color);
        
        // Draw simple face
        screen.fillCircle(x - 5 * scale, y - 3 * scale, 2 * scale, { r: 0, g: 0, b: 0, a: 255 });
        screen.fillCircle(x + 5 * scale, y - 3 * scale, 2 * scale, { r: 0, g: 0, b: 0, a: 255 });
    }
    
    getAvatarColor(avatarType) {
        // Simple color mapping for avatars
        const colors = [
            { r: 255, g: 200, b: 150, a: 255 }, // Human - peach
            { r: 139, g: 90, b: 43, a: 255 },   // Dog - brown
            { r: 255, g: 165, b: 0, a: 255 },   // Cat - orange
            { r: 200, g: 200, b: 200, a: 255 }, // Rabbit - gray
            { r: 255, g: 220, b: 100, a: 255 }, // Giraffe - yellow
            { r: 139, g: 69, b: 19, a: 255 },   // Owl - dark brown
            { r: 100, g: 50, b: 0, a: 255 },    // Millipede - dark brown
            { r: 210, g: 180, b: 140, a: 255 }, // Puppy - tan
            { r: 255, g: 140, b: 0, a: 255 },   // Tiger - orange
            { r: 101, g: 67, b: 33, a: 255 },   // Chimpanzee - dark brown
            { r: 139, g: 90, b: 43, a: 255 },   // Platypus - brown
            { r: 205, g: 133, b: 63, a: 255 },  // Lynx - tan
            { r: 0, g: 100, b: 0, a: 255 },     // Gator - dark green
            { r: 255, g: 200, b: 100, a: 255 }, // Ocelot - light orange
            { r: 255, g: 255, b: 200, a: 255 }  // Hen - light yellow
        ];
        
        return colors[avatarType] || colors[0];
    }
}